src/Mapper.cpp

```diff
@@ -1346,3 +1346,223 @@ map<int, CGRANode*>* Mapper::getReorderPath(map<CGRANode*, int>* t_path) {
   return reorderPath;
 }
 
+
+// save the mapping results to json file for subsequent incremental mapping
+void Mapper::generateJSON4IncrementalMap(CGRA* t_cgra, DFG* t_dfg){
+        ofstream jsonFile("increMapInput.json", ios::out);
+        jsonFile<<"{"<<endl;
+        jsonFile<<"     \"Opt2TileXY\":{"<<endl;
+        int idx = 0;
+        for (DFGNode* dfgNode: t_dfg->nodes) {
+                // write dfgnodeID, mapped CGRANode X and Y coordinates
+                jsonFile<<"             \""<<dfgNode->getID()<<"\": {"<<endl; // opt id
+                jsonFile<<"                     \"x\":"<<m_mapping[dfgNode]->getX()<<","<<endl; // opt mapped tile x coordinate
+                jsonFile<<"                     \"y\":"<<m_mapping[dfgNode]->getY()<<endl; // opt mapped tile y coordinate
+                idx += 1;
+                if(idx < t_dfg->nodes.size()) jsonFile<<"             },"<<endl;
+                else jsonFile<<"        }"<<endl;
+        }
+        jsonFile<<"     },"<<endl;
+
+        jsonFile<<"     \"Tile2Level\":{"<<endl;
+        // generate level informations of current mapping results
+        map<int, vector<CGRANode*>> FanIO_CGRANodes;
```

Comment:
Can you plz add comments about what is FanIO to the other readers? And why inport/outport would be associated with a vector of `CGRANodes`?

Reply:
FanIO is the number of links of current CGRANode connected to other CGRANode.
For example, in a 4x4 CGRA with mesh topology
0    1    2    3
4    5    6    7
8    9    10  11
12  13  14  15
CGRANode5 is connected to CGRANode1, 4, 6, and 9, therefore it has FanI/O = 4
CGRANode1 is connected to CGRANode0, 2, and 5, therefore it has FanI/O = 3
CGRANode0 is connected to CGRANode1, and 4, therefore it has FanI/O = 2

Then we have
map<int, vector<CGRANode*>> FanIO_CGRANodes = {
    {4, {CGRANode5, CGRANode6, CGRANode9, CGRANode10}},
    {3, {CGRANode1, CGRANode2, CGRANode4, CGRANode7, CGRANode8, CGRANode11, CGRANode13, CGRANode14}},
    {2, {CGRANode0, CGRANode3, CGRANode12, CGRANode15}}
}
In incremental mapping, FanIO_CGRANodes can help with querying the list of CGRANodes with the given FanIO

Reply:
Plz put these as comments above the `FanIO_CGRANodes`. Thanks!

Reply:
Has been corrected